### PR TITLE
[Feature] Add TR3D detector to projects

### PR DESCRIPTION
**Pull request by filaPro**
## Modification

Add TR3D detector in ScanNet, SUN RGB-D and S3DIS to `projects/TR3D`.

## BC-breaking (Optional)

No

### CODE REVIEW

File: projects/TR3D/tr3d/tr3d_neck.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    def _init_layers(self, in_channels: Tuple[int], out_channels: int):
+        """Initialize layers.
+
+        Args:
+            in_channels (tuple[int]): Number of channels in input tensors.
+            out_channels (int): Number of channels in output tensors.
+        """
+        for i in range(len(in_channels)):
+            if i > 0:
+                self.__setattr__(
```
Comment (ZwwWayne) [inline projects/TR3D/tr3d/tr3d_neck.py:41]: can we use self.add_module here?

**Comment by ZwwWayne** [comment]
Hi @filaPro ,
Thanks for your kind contribution. Overall, this PR looks good. We can merge this PR after fixing small comments.

File: projects/TR3D/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# TR3D: Towards Real-Time Indoor 3D Object Detection
```
Comment (ZwwWayne) [inline projects/TR3D/README.md:1]: The contents in this README.md looks good.
To let users better understand its status and usages in projects, we also recommend developers to follow this [template](https://github.com/open-mmlab/mmdetection3d/blob/dev-1.x/projects/example_project/README.md) to provide more information.

File: projects/TR3D/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  title={TR3D: Towards Real-Time Indoor 3D Object Detection},
+  author={Rukhovich, Danila and Vorontsova, Anna and Konushin, Anton},
+  journal={arXiv preprint arXiv:2302.02858},
+  year={2023}
+}
+```
+
+## Checklist
+
+- \[✓\] Milestone 1: PR-ready, and acceptable to be one of the `projects/`.
```
Comment (Xiangxu-0103) [inline projects/TR3D/README.md:67]: Suggest to use `- [x]` instead of `- \[✓\]` for markdown format.

File: projects/TR3D/configs/tr3d.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+_base_ = [
+    '../../../configs/_base_/default_runtime.py',
```
Comment (Xiangxu-0103) [inline projects/TR3D/configs/tr3d.py:2]:
```suggestion
    'mmdet3d::_base_/default_runtime.py',
```

File: projects/TR3D/configs/tr3d_1xb16_s3dis-3d-5class.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+_base_ = ['./tr3d.py', '../../../configs/_base_/datasets/s3dis-3d.py']
```
Comment (Xiangxu-0103) [inline projects/TR3D/configs/tr3d_1xb16_s3dis-3d-5class.py:1]:
```suggestion
_base_ = ['./tr3d.py', 'mmdet3d::_base_/datasets/s3dis-3d.py']
```

File: projects/TR3D/configs/tr3d_1xb16_scannet-3d-18class.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+_base_ = ['./tr3d.py', '../../../configs/_base_/datasets/scannet-3d.py']
```
Comment (Xiangxu-0103) [inline projects/TR3D/configs/tr3d_1xb16_scannet-3d-18class.py:1]: The same as above.

File: projects/TR3D/configs/tr3d_1xb16_sunrgbd-3d-10class.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
+_base_ = ['./tr3d.py', '../../../configs/_base_/datasets/sunrgbd-3d.py']
```
Comment (Xiangxu-0103) [inline projects/TR3D/configs/tr3d_1xb16_sunrgbd-3d-10class.py:1]: The same as above.

File: projects/TR3D/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
```
Comment (Xiangxu-0103) [inline projects/TR3D/README.md:8]: Better to add an illustration here.

File: projects/TR3D/configs/tr3d_1xb16_s3dis-3d-5class.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        keys=['points', 'gt_bboxes_3d', 'gt_labels_3d'])
+]
+
+train_dataloader = dict(
+    batch_size=16,
+    num_workers=8,
+    sampler=dict(type='DefaultSampler', shuffle=True),
+    dataset=dict(
+        type='RepeatDataset',
+        # times=13,
```
Comment (Xiangxu-0103) [inline projects/TR3D/configs/tr3d_1xb16_s3dis-3d-5class.py:44]: Please remove the comment here, and also check others.

File: projects/TR3D/tr3d/rotated_iou_loss.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    zmax1 = box3d1[..., 2] + box3d1[..., 5] * 0.5
+    zmin1 = box3d1[..., 2] - box3d1[..., 5] * 0.5
+    zmax2 = box3d2[..., 2] + box3d2[..., 5] * 0.5
+    zmin2 = box3d2[..., 2] - box3d2[..., 5] * 0.5
+    z_overlap = (torch.min(zmax1, zmax2) -
+                 torch.max(zmin1, zmin2)).clamp_(min=0.)
+    intersection_3d = intersection * z_overlap
+    volume1 = box3d1[..., 3] * box3d1[..., 4] * box3d1[..., 5]
+    volume2 = box3d2[..., 3] * box3d2[..., 4] * box3d2[..., 5]
+    union_3d = volume1 + volume2 - intersection_3d
```
Comment (Xiangxu-0103) [inline projects/TR3D/tr3d/rotated_iou_loss.py:37]: Line 23-37 seems have been implemented in `diff_iou_rotated_3d` [here](https://github.com/open-mmlab/mmcv/blob/2.x/mmcv/ops/diff_iou_rotated.py#L276). We may directly use this function for simplify.

File: projects/TR3D/tr3d/tr3d_head.py [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+            torch.sqrt(
+                torch.pow(bbox_pred[:, 6], 2) + torch.pow(bbox_pred[:, 7], 2)))
+        alpha = 0.5 * torch.atan2(bbox_pred[:, 6], bbox_pred[:, 7])
+        return torch.stack(
+            (x_center, y_center, z_center, scale / (1 + q), scale /
+             (1 + q) * q, bbox_pred[:, 5] + bbox_pred[:, 4], alpha),
+            dim=-1)
+
+    @torch.no_grad()
+    def _get_targets(self, points: Tensor, gt_bboxes: BaseInstance3DBoxes,
```
Comment (Xiangxu-0103) [inline projects/TR3D/tr3d/tr3d_head.py:357]: _get_targets -> get_targets

File: projects/TR3D/tr3d/transforms_3d.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+        Returns:
+            tuple[:obj:`BasePoints`, np.ndarray] | :obj:`BasePoints`:
+
+                - points (:obj:`BasePoints`): 3D Points.
+                - choices (np.ndarray, optional): The generated random samples.
+        """
+        if type(self.num_points) is float:
+            num_samples = int(
+                np.random.uniform(self.num_points, 1.) * points.shape[0])
```
Comment (Xiangxu-0103) [inline projects/TR3D/tr3d/transforms_3d.py:49]:
```suggestion
        if isinstance(num_samples, float):
            assert num_samples < 1
            num_samples = int(
                np.random.uniform(num_samples, 1.) * points.shape[0])
```

**Comment by Xiangxu-0103** [comment]
Overall, this PR LGTM.

File: projects/TR3D/tr3d/rotated_iou_loss.py (hunk 10 above)
Comment (filaPro) [reply projects/TR3D/tr3d/rotated_iou_loss.py:37]: But we need lines 23-26 and 28-31 after 37 line

Comment (Xiangxu-0103) [reply projects/TR3D/tr3d/rotated_iou_loss.py:37]: Yeah, you are right. It's my fault. Thanks for your reminder.

**Review by Xiangxu-0103 — APPROVED** [review]

**Comment by ZwwWayne** [comment]
Hi @filaPro ,
This PR is OK to be merged. Would you like to update the links of models, logs, and their log.json? It is fine if you provide a way for us to download and upload them to our host, then we can send the links to you for update.

**Comment by filaPro** [comment]
Hi @ZwwWayne ,
Can you please upload `tr3d_*` from [here](https://github.com/filaPro/tmp/releases/tag/v0.1) and provide the links?

**Comment by Tai-Wang** [comment]
Hi @filaPro , I have uploaded the checkpoints and logs to our server. You can update the links following the template "https://download.openmmlab.com/mmdetection3d/v1.1.0_models/tr3d/xxx/xxx.log.json", like "https://download.openmmlab.com/mmdetection3d/v1.1.0_models/tr3d/tr3d_1xb16_scannet-3d-18class/tr3d_1xb16_scannet-3d-18class.log.json".

**Comment by filaPro** [comment]
Hi @Tai-Wang ,
Thanks a lot.
But i followed `projects` directory where all config paths are `PETR/petr` all `CenterFormer/centerformer`. Should i rename `TR3D/tr3d` to `tr3d/tr3d`?

**Comment by Tai-Wang** [comment]
> Hi @Tai-Wang , Thanks a lot. But i followed `projects` directory where all config paths are like `PETR/petr` or `CenterFormer/centerformer`. Should i rename `TR3D/tr3d` to `tr3d/tr3d`?

Sorry, the current name should be ok in projects. Only need to name the config folder name to "tr3d" when we integrate these code into the core codes of mmdet3d.

**Review by Tai-Wang — APPROVED** [review]

**Comment by ZCMax** [comment]
Thanks for your PR, it LGTM for me to be merged into projects
